[2단계 - 로또 리팩토링] 엠제이(최민준) 미션 제출합니다.

theminjunchoi opened the pull request:
안녕하세요 앤지! 
벌써 주말이 끝나가네요 시간이 너무 빠른 것 같아요😭
지난 번에 이어서 리팩토링 한 내용들을 정리해볼게요.

## 리팩토링한 내용
우선 1단계에서 `LottoNumber` 일급 객체를 도입하면서 많은 부분을 리팩토링을 했기 때문에 크게 달라진 부분은 없어요.
대신, `WinningResponse` dto를 `WinningResult` model로 바꾸면서, 
당첨 결과를 계산하는 역할을 `WinningLotto`에서 `WinningResult`로 옮겨주었습니다.

생각을 해봤는데, 당첨로또는 로또와 보너스번호에 대한 로직만 있으면 되고, 다른 로또와 비교해서 당첨 결과를 알려주는 로직은  `WinningResult`가 생겼는데 굳이 `WinningLotto`가 갖고 있는게 어색하다고 느껴졌어요.

## 궁금한 내용
view 계층에서 입력에 대한 validation까지해서 controller로 보내주는 건 어떻게 생각하시나요?
저는 view가 validation까지해서 넘겨주는 것과, controller에서 일단 받아서 validation 해주는 것 둘 다 괜찮다고 생각이 드는데,
굳이 뽑자면 전자의 방식이 controller의 부담을 줄여줄 수 있을 것 같아서 좋아보이긴합니다!
우선 이 서비스에서는 후자로 처리를 했어요! 앤지의 의견이 궁금합니다~

추가로 수정 및 개선이 필요한 부분이 있다면 알려주세요!
시간내어 리뷰해주셔서 감사합니다~☺️

theminjunchoi committed:
docs: 기능 요구사항 작성

theminjunchoi committed:
feat: 구입금액 입력받는 기능 구현

ykmxxi committed:
feat: 로또 개수 계산하는 기능 구현

theminjunchoi committed:
feat: 로또 번호 예외처리 기능 구현

ykmxxi committed:
feat: 숫자 6개 자동생성하는 기능 구현

theminjunchoi committed:
feat: 로또개수에 맞게 발급하는 기능 구현

ykmxxi committed:
feat: 당첨 로또를 생성하는 기능 구현

theminjunchoi committed:
test: Lotto 테스트 코드 추가

ykmxxi committed:
feat: 로또 등수를 계산하는 기능 구현

theminjunchoi committed:
feat: 로또 당첨 결과 출력하는 기능 구현

ykmxxi committed:
feat: 로또 당첨 결과를 계산하는 기능 구현

theminjunchoi committed:
feat: 로또 당첨 수익률 계산하는 기능 구현

ykmxxi committed:
feat: 발행된 로또를 출력하는 기능 구현

ykmxxi committed:
refactor: 일관된 컨벤셕 적용

ykmxxi committed:
test: 당첨 등수 계산 기능 반환 타입 수정

ykmxxi committed:
fix: 잘모된 파라미터 입력 수정

ykmxxi committed:
refactor: 컬렉션 자료형을 상위 인터페이스로 수정

ykmxxi committed:
refactor: 로또 숫자 범위 검증 메서드 분리

ykmxxi committed:
refactor: 매칭된 로또 번호 계산 기능 스트림 적용

ykmxxi committed:
refactor: 로또 컨트롤러 실행 메서드 분리

theminjunchoi committed:
refactor: 출력 메소드 분리

theminjunchoi committed:
refactor: 로또 당첨 결과 저장 메소드 분리

theminjunchoi committed:
refactor: 인스턴스화 막는 private 생성자 추가

ykmxxi committed:
docs: README.md 갱신

ykmxxi committed:
remove: 불필요한 .gitkeep 파일 삭제

theminjunchoi committed:
refactor: 테스트 코드 메소드명 변경

theminjunchoi committed:
fix: 파라미터에 final 키워드 추가

theminjunchoi committed:
fix: 랜덤 번호 생성하는 random 객체를 static 인스턴스 변수로 수정

theminjunchoi committed:
refactor: LottoNumber 일급 객체 사용하여 리팩토링

theminjunchoi committed:
fix: 안쓰는 코드 제거

theminjunchoi committed:
comment: 테스트 코드 주석 해제

theminjunchoi committed:
style: 돈 세자리 마다 언더바(_) 넣어주기

theminjunchoi committed:
fix: 구입 금액 읽는 메소드명 변경

theminjunchoi committed:
fix: 구입 금액에 음수를 입력할 때 에러를 발생하도록 수정

theminjunchoi committed:
fix: 당첨 로또에 중복된 숫자를 입력한 경우, 보너스 번호 입력 전 에러문구 출력하도록 수정

theminjunchoi committed:
rename: WinningResponse를 WinnningResult로 변경

theminjunchoi committed:
rename: hasNumber 메소드 이름 변경

theminjunchoi committed:
refactor: Money 클래스로 구매금액 및 수익률 구하는 기능 구현

theminjunchoi committed:
refactor: 안 쓰는 코드 제거

theminjunchoi committed:
style: 상수 변수명 변경

theminjunchoi committed:
docs: 기존 문제점들 정리

theminjunchoi committed:
refactor: 테스트 메소드명 변경

theminjunchoi committed:
test: Money 입력 관련 테스트 추가

theminjunchoi committed:
refactor: WinningLotto가 WinningResult를 모르도록 수정

theminjunchoi committed:
refactor: 채점 시작하는 메소드명 변경

theminjunchoi committed:
docs: 리팩토링 내용 정리

theminjunchoi committed:
Merge branch 'theminjunchoi' into step2

soominsohn commented on src/main/java/lotto/model/WinningResults.java:
```diff
@@ -1,25 +1,80 @@
 
-    public List<WinningResult> getResponses() {
-        return responses;
+    private static void saveResults(final Rank rank, final boolean hasBonus, final Map<Rank, Integer> ranks,
```
```suggestion
    private void saveResults(final Rank rank, final boolean hasBonus, final Map<Rank, Integer> ranks,
```
불필요한 static은 제거해보면 어떨까요?

soominsohn commented on src/main/java/lotto/model/WinningResults.java:
```diff
@@ -1,25 +1,80 @@
+    private final List<WinningResult> results;
 
-    private final List<WinningResult> responses;
+    public WinningResults(WinningLotto winningLotto, Lottos lottos) {
+        this.winningLotto = winningLotto;
+        this.lottos = lottos;
+        Map<Rank, Integer> ranks = scoreRanks();
+        this.results = toResults(ranks);
+    }
```
생성자에서 scoreRanks() 메서드를 호출하고, 그 결과를 가공하여 results 필드를 초기화하고 있군요..!
생성자의 책임이 커지고, 테스트가 어려워질 수 있습니다.
생성자는 객체의 필드를 초기화하는 역할만 수행하도록 만들고 별도의 메서드를 만들어 로직을 수행하는 것이 좋습니다.

theminjunchoi replied:
아하 그렇군요! 
이것도 반영해서 수정해볼게요!

soominsohn commented on src/main/java/lotto/model/WinningResults.java:
```diff
@@ -1,25 +1,80 @@
+    }
+
+    private Map<Rank, Integer> scoreRanks() {
+        Map<Rank, Integer> ranks = new LinkedHashMap<>();
+        initRanks(ranks);
+        saveRanks(ranks);
```
메서드에서 매개변수 ranks를 인자로 넣어주고 이를 직접 변경하고 있군요. 이는 ranks가 어디서 변경됐는지 파악하기 쉽지 않으며 디버깅을 어렵게 만들 수 있습니다.
또 initRanks가 단순히 Rank를 키로 가지고 있는 Map을 초기화하는 역할을 하고 있는데요, saveRanks 메서드 에서 호출하는 것도 괜찮아 보입니다.
```suggestion
        Map<Rank, Integer> savedRanks = initRanks(ranks);
```
함수들의 return값으로 변경된 ranks를 반환하게 명시적으로 표현해보는건 어떨까요?

theminjunchoi replied:
좋은 의견인 것 같습니다!
수정해볼게요!

soominsohn commented on src/main/java/lotto/model/WinningResults.java:
```diff
@@ -1,25 +1,80 @@
         long totalEarnedMoney = 0L;
-        for (WinningResult response : this.getResponses()) {
-            totalEarnedMoney += response.getWinningAmount() * response.getWinningCount();
+        for (WinningResult result : this.getResults()) {
+            totalEarnedMoney += result.getWinningAmount() * result.getWinningCount();
```
`result.getWinningAmount() * result.getWinningCount();` 이 계산하는 로직을 WinningResult의 메서드로 옮겨볼 수 있을 것 같아요. [이 글](https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/)을 참고해보시면 좋습니다.

theminjunchoi replied:
오 `Collections.unmodifiableList()`를 사용하는 걸 처음 알았네요!
수정해보겠습니다~!

soominsohn commented on src/main/java/lotto/model/WinningResults.java:
```diff
@@ -1,25 +1,80 @@
 
+
     public long calculateEarnedMoney() {
         long totalEarnedMoney = 0L;
-        for (WinningResult response : this.getResponses()) {
-            totalEarnedMoney += response.getWinningAmount() * response.getWinningCount();
+        for (WinningResult result : this.getResults()) {
```
`WinningResults` 객체에 대한 테스트도 작성해볼까요?

theminjunchoi replied:
우선 작성을 하고 push를 했는데 제가 생각했을 땐 조금 더 효율적인 테스트 코드를 작성해야할 것 같아요.
혹시 확인하고 수정 및 개선 사항을 더 알려줄 수 있을까요?

soominsohn replied:
WinningResultsTest에 달아놨어요~
한번 읽어보시고 다음미션 때 천천히 적용해보세요!
질문있으시면 언제든지 말씀해주세요 😄

soominsohn commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -27,11 +26,12 @@ public LottoController(final InputView inputView, final OutputView outputView) {
         try {
-            Money money = new Money(Integer.parseInt(inputView.readPurchaseMoney()));
+
+            Money money = new Money(parseInt(inputView.readPurchaseMoney()));
```
주신 Controller vs View 검증 질문에 대한 답변입니다!

View 계층에서 입력을 검증하면 초기단계에서 잘못된 입력을 걸러내어 Controller로 전달되는 부적절한 데이터를 줄일 수 있습니다. 
그러나 입력 채널이 변경된다면 각 View마다 검증 로직을 수정해야하는 단점이 생기겠네요.
전 두 방식 다 괜찮다고 생각이 들기때문에 엠제이가 취향대로 선택하시면 될 것 같아요. 😄 
[참고](https://velog.io/@pengoose_dev/MVC-Pattern%EC%97%90%EC%84%9C-%EC%9C%A0%ED%9A%A8%EC%84%B1-%EA%B2%80%EC%82%AC%EB%8A%94-%EB%88%84%EA%B5%AC%EC%9D%98-%EC%B1%85%EC%9E%84%EC%9D%B8%EA%B0%80)

또, 이전 리뷰에서 못드린 리뷰가 있는데
Lotto 객체의 validateDuplication 메서드가 `규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.`를 지키고 있지 않아요. 개선해보면 좋을 것 같아요. 👍

theminjunchoi replied:
저도 저 글을 읽어봤는데, 아직까지는 두 방식 다 괜찮아보여요!
아마 다른 미션들을 하면서 제 생각의 방향성이 잡힐 것 같습니다!

우선 인덴트 관련사항은 수정했습니다!

soominsohn reviewed (changes requested):
안녕하세요 엠제이~ 2단계 잘 봤습니다. 😄 
코멘트 몇개 남겼으니 확인해주시고 궁금한점 있으시면 질문주세요~!

theminjunchoi committed:
refactor: 불필요한 static 제거

theminjunchoi committed:
Merge remote-tracking branch 'origin/step2' into step2

theminjunchoi committed:
refactor: ranks를 초기화하여 채점하는 메소드 수정
- initRanks()를 saveRanks에서 호출하도록 수정
- initRanks(), saveRanks() 메소드에서 ranks를 반환하도록 수정

theminjunchoi committed:
refactor: WinningResults의 생성자가 초기화만 할 수 있도록 수정

theminjunchoi committed:
refactor: Rank별 당첨 금액 계산하는 메소드를 WinningResult로 분리

theminjunchoi committed:
fix: indent 2개인 메소드 수정

theminjunchoi committed:
test: WinningResults 테스트 코드 작성

soominsohn commented on src/main/java/lotto/model/WinningResult.java:
```diff
@@ -14,6 +14,10 @@ public WinningResult(int matchingCount, long winningAmount, boolean hasBonus, in
         this.winningCount = winningCount;
     }
 
+    public long calculateWinningMoney() {
+        return winningAmount * winningCount;
+    }
+
```
잘 반영해주셨네요 😄

soominsohn commented on src/test/java/lotto/model/WinningResultsTest.java:
```diff
@@ -0,0 +1,57 @@
+        }
+        WinningResults winningResults = new WinningResults(winningLotto, lottos);
+        winningResults.calculateWinningResults();
+        Assertions.assertThat(money.calculateReturnRatio(winningResults.calculateEarnedMoney())).isEqualTo(16.66);
```
```suggestion
        assertThat(winningResults.calculateEarnedMoney()).isEqualTo(XXXXX);
```
money가 earningRatio를 계산하는건 MoneyTest에서 이뤄졌으니 winningResults가 벌어들인 금액을 검증하면 될 것 같아요!
WinningResults 객체에 대한 단위테스트니까요!

theminjunchoi replied:
앗 그러네요!

soominsohn commented on src/test/java/lotto/model/WinningResultsTest.java:
```diff
@@ -0,0 +1,57 @@
+    @MethodSource("provideLottos")
+    @ParameterizedTest
+    void calculateReturnRatio(List<Lotto> provideLottos) {
+        Money money = new Money(3_000);
+        Lotto lotto = new Lotto(
+                List.of(1, 2, 3, 4, 5, 6).stream()
+                        .map(LottoNumber::new)
+                        .collect(Collectors.toList())
+        );
+        LottoNumber bonusNumber = new LottoNumber(7);
```
테스트에서 사용되는 데이터들은 따로 테스트픽스쳐라는 것을 만들어서 관리해주면 좋아요.
Money, Lotto, LottoNumber 모두 테스트 전반에서 사용될텐데 테스트픽스쳐로 관리해보면 좋습니다.
다음 미션하시면서 적용해보세용 ㅎㅎ!

soominsohn commented on src/test/java/lotto/model/WinningResultsTest.java:
```diff
@@ -0,0 +1,57 @@
+            lottos.add(provideLotto);
+        }
+        WinningResults winningResults = new WinningResults(winningLotto, lottos);
+        winningResults.calculateWinningResults();
+        Assertions.assertThat(money.calculateReturnRatio(winningResults.calculateEarnedMoney())).isEqualTo(16.66);
```
`winningResults.calculateWinningResults();`
`winningResults.calculateEarnedMoney();`
두개의 메서드의 상호작용을 검증하고 있네요.
클래스의 메서드를 작성할 때 `메서드는 한가지 일을 해야한다.`라는 규칙을 지키면서 작성한 것처럼
테스트 메서드도 한가지 행위에 대한 결과를 검증하면 좋아요.

WinningResults의 메서드들이 하나하나 잘 동작하는지 테스트 코드를 작성해주면 좋습니다.
테스트코드를 작성하시다보면 고민이 또 생길텐데, 그 때 질문주셔도 좋습니다 😄

soominsohn reviewed (approved):
안녕하세요 엠제이~
2단계 정말 수고많으셨습니다.
첫 미션인데 정말 잘해주셨고 다음 미션도 잘하실거에요~!
파이팅입니다 😄
